File: OC_XFetchCodeButton/Resource/XFetchCodeButton.m

//
//  XFetchCodeButton.m
//  XFetchCodeButton
//
//  Created by 王雄皓 on 2017/4/19.
//  Copyright © 2017年 王雄皓. All rights reserved.
//

#import "XFetchCodeButton.h"

@implementation XFetchCodeButton
{
    NSTimer *_timer; // 定时器
    UIButton *_codeBtn; // 倒计时按钮
    
    int _timerCount;
    id _pBlock;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupUI];
    }
    return self;
}

- (void)setNormalTitle:(NSString *)normalTitle {
    
    _normalTitle = normalTitle;
    [_codeBtn setTitle:_normalTitle forState:UIControlStateNormal];
}

- (void)setNormalTintColor:(UIColor *)normalTintColor {
    
    _normalTintColor = normalTintColor;
    [_codeBtn setTitleColor:_normalTintColor forState:UIControlStateNormal];
}

- (void)setDisabledColor:(UIColor *)disabledColor {
    
    _disabledColor = disabledColor;
    [_codeBtn setTitleColor:_disabledColor forState:UIControlStateDisabled];
}


- (void)setFontSize:(CGFloat)fontSize {
    
    _fontSize = fontSize;
    [_codeBtn.titleLabel setFont:[UIFont systemFontOfSize:_fontSize]];
}


- (void)setupUI {
    
    self.normalTitle = @"获取验证码";
    self.normalTintColor = [UIColor blueColor];
    self.fontSize = 10;
    self.disabledColor = [UIColor grayColor];
    _timerCount = CountDownDefault;
    
    _codeBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [_codeBtn setFrame:CGRectMake(0, 0, self.frame.size.width, self.frame.size.height)];
    [_codeBtn setTitle:self.normalTitle forState:UIControlStateNormal];
    [_codeBtn setTitleColor:self.normalTintColor forState:UIControlStateNormal];
    [_codeBtn setTitleColor:self.disabledColor forState:UIControlStateDisabled];
    [_codeBtn setBackgroundColor:[UIColor clearColor]];
    [_codeBtn.titleLabel setFont:[UIFont systemFontOfSize:self.fontSize]];
    [_codeBtn addTarget:self action:@selector(send) forControlEvents:UIControlEventTouchUpInside];
    [self setBackgroundColor:[UIColor clearColor]];
    [self addSubview:_codeBtn];
    
}

- (void)addTarget:(id)target action:(SEL)sel forEvents:(UIControlEvents)events {
    
    [_codeBtn addTarget:target action:sel forControlEvents:events];
}

- (void)doneSend:(void (^)())block {
    
    _pBlock = block;
}

- (void)send {
    
    if (_pBlock) {
        ((void(^)())_pBlock)();
    }
    
    [_codeBtn setEnabled:false];
    _timer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(start:) userInfo:nil repeats:true];
    
}

- (void)stop {
    
    if (_timer) {
        [_timer invalidate];
        _timer = nil;
        _timerCount = CountDownDefault;
    }
}

- (void)start:(NSTimer *)ti {
    __weak typeof(self) __weakSelf = self;

    if (_timerCount == 1) {
        
        if (_timer) {
            [_timer invalidate];
            _timer = nil;
        }
        
        _timerCount = CountDownDefault;
        dispatch_async(dispatch_get_main_queue(), ^{
            [_codeBtn setEnabled:true];
            [_codeBtn setTitle:__weakSelf.normalTitle forState:UIControlStateNormal];
        });
        return;
    }
    
    _timerCount -= 1;
    NSString * old = [CountDown copy];
    NSString * new = [[[NSString alloc] initWithFormat:@"%d", _timerCount] stringByAppendingString:old];

    dispatch_async(dispatch_get_main_queue(), ^{
        [_codeBtn setTitle:new forState:UIControlStateNormal];
    });
}

@end

File: XFetchCodeButton/Resource/XFetchCodeButton.swift
//
//  XFetchCodeButton
//  XFetchCodeButton
//
//  Created by 王雄皓 on 2017/4/19.
//  Copyright © 2017年 王雄皓. All rights reserved.
//
import UIKit


public class XFetchCodeButton: UIView {
    
    static let countDownDefault: Int = 10 // 倒计时时间

    fileprivate var timer: Timer? // 定时器
    fileprivate var codeBtn: UIButton? // 倒计时按钮
    fileprivate var doneSendBlock: (() -> Void)? // 回调块

    fileprivate lazy var timerCount: Int = countDownDefault
    fileprivate lazy var countDown: String = "秒后重发" // 默认后缀
    
    var disabledColor: UIColor = UIColor.gray { //默认不可用状态颜色
        
        didSet {
            
            guard let btn = codeBtn else { return }
            btn.setTitleColor(disabledColor, for: .disabled)
        }
    }
    
    var fontSize: CGFloat = 10 {// 默认字体大小
        
        didSet {
            
            guard let btn = codeBtn else { return }
            btn.titleLabel?.font = UIFont.systemFont(ofSize: fontSize)
        }
    }
    
    var normalTintColor: UIColor = UIColor.blue { // 默认字体颜色
        
        didSet {
            setTitleColor(normalTintColor)
        }
    }
    
    public var normalTitle: String = "获取验证码" { // 默认normal显示文字
       
        didSet {
            setTitle(normalTitle)
        }
    }
    
    override init(frame: CGRect) {
        
        super.init(frame: frame)
        
        codeBtn = {
            let button = UIButton(type: .custom)
            button.setTitle(normalTitle, for: .normal)
            button.setTitleColor(normalTintColor, for: .normal)
            button.titleLabel?.font = UIFont.systemFont(ofSize: fontSize)
            button.frame = CGRect(x: 0, y: 0, width: frame.size.width, height: frame.size.height)
            button.setTitleColor(disabledColor, for: .disabled)
            button.backgroundColor = .clear
            button.addTarget(self, action: #selector(codeSend), for: .touchUpInside)
            return button
        }()
        self.backgroundColor = .clear
        addSubview(codeBtn!)
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func layoutSubviews() { // 重写codebtn的frame
        super.layoutSubviews()
        codeBtn!.frame = CGRect(x: 0, y: 0, width: frame.size.width, height: frame.size.height)
    }
    
    /// target 添加
    public func addTarget(target: Any?, action: Selector, forEvents: UIControlEvents) {
        
        guard let btn = codeBtn else { return }
        btn.addTarget(target, action: action, for: forEvents)
    }

    /// 点击回调
    public func doneSend(_ callBack: @escaping () -> Void) {
        
        doneSendBlock = callBack
    }
    
}
fileprivate extension XFetchCodeButton {
    
    /// 设置未点击文字颜色
    func setTitleColor(_ forNormal: UIColor) {
        
        guard let btn = codeBtn else { return }
        btn.setTitleColor(forNormal, for: .normal)
    }
    
    /// 设置未点击文字
    func setTitle(_ forNormal: String) {
        
        guard let btn = codeBtn else { return }
        btn.setTitle(forNormal, for: .normal)
    }
    
    @objc func codeSend() {
        
        guard let btn = codeBtn else { return }
        
        if let block = doneSendBlock {
            block()
        }
        
        btn.isEnabled = false
        timer = Timer.scheduledTimer(timeInterval: TimeInterval(1), target: self, selector: #selector(self.start(_:)), userInfo: nil, repeats: true)
    }
    
    
    func stop() {
        
        if let ti = timer {
            ti.invalidate()
            timer = nil
            timerCount = XFetchCodeButton.countDownDefault
        }
    }
    
    @objc func start(_ time: Timer) {
        
        if timerCount == 1 {
            
            if let ti = timer {
                ti.invalidate()
                timer = nil
            }
            
            timerCount = XFetchCodeButton.countDownDefault
            DispatchQueue.main.async {
                if let button = self.codeBtn {
                    button.isEnabled = true
                    button.setTitle(self.normalTitle, for: .normal)
                }
            }
            return
        }
        
        timerCount -= 1
        DispatchQueue.main.async {
            if let button = self.codeBtn {
                button.setTitle(String(self.timerCount) + self.countDown, for: .normal)
            }
        }
    }
    
}
